services/paymaster/src/constraint.rs

```diff
@@ -0,0 +1,244 @@
+use serde::{Deserialize, Serialize};
+use serde_with::{serde_as, DisplayFromStr};
+use solana_pubkey::Pubkey;
+use utoipa::ToSchema;
+
+use crate::utils::deserialize_pubkey_vec;
+
+#[derive(Serialize, Deserialize, ToSchema)]
+#[serde(tag = "version")]
+pub enum TransactionVariation {
+    #[serde(rename = "v0")]
+    #[schema(title = "v0")]
+    V0(VariationProgramWhitelist),
+
+    #[serde(rename = "v1")]
+    #[schema(title = "v1")]
+    V1(VariationOrderedInstructionConstraints),
+}
+
+#[derive(Serialize, Deserialize, ToSchema)]
+#[serde_as]
+pub struct VariationProgramWhitelist {
+    pub name: String,
+
+    #[schema(example = "[\"So11111111111111111111111111111111111111111\"]", value_type = Vec<String>)]
+    #[serde(deserialize_with = "deserialize_pubkey_vec")]
+    pub whitelisted_programs: Vec<Pubkey>,
+}
+
+#[derive(Serialize, Deserialize, ToSchema)]
+pub struct VariationOrderedInstructionConstraints {
+    pub name: String,
+    pub instructions: Vec<InstructionConstraint>,
+    pub rate_limits: RateLimits,
+    pub max_gas_spend: u64,
+}
+
+#[derive(Serialize, Deserialize, ToSchema)]
+pub struct RateLimits {
+    user_per_min: Option<u64>,
+    session_per_min: Option<u64>,
+    ip_per_min: Option<u64>,
+}
+
+#[serde_as]
+#[derive(Serialize, Deserialize, ToSchema)]
+pub struct InstructionConstraint {
+    #[schema(example = "So11111111111111111111111111111111111111111", value_type = String)]
+    #[serde_as(as = "DisplayFromStr")]
+    pub program: Pubkey,
+    pub accounts: Vec<AccountConstraint>,
+    pub data: Vec<DataConstraint>,
+    pub required: bool,
+}
+
+#[derive(Serialize, Deserialize, ToSchema)]
+pub struct AccountConstraint {
+    pub index: u16,
+    pub include: Vec<ContextualPubkey>,
+    pub exclude: Vec<ContextualPubkey>,
+}
+
+#[serde_as]
+#[derive(Serialize, Deserialize, ToSchema)]
+pub enum ContextualPubkey {
+    Explicit {
+        #[schema(example = "So11111111111111111111111111111111111111111", value_type = String)]
+        #[serde_as(as = "DisplayFromStr")]
+        pubkey: Pubkey,
+    },
+    Sponsor,
+    Signer {
+        index: i8,
+    },
+    UserDelegator,
+}
+
+pub trait ContextualPubkeyTrait {
```

Comment:
Minor nit but I don't think this trait name tells me much about what the trait actually does